src/v/http/rest_client/rest_entity.cc

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "http/rest_client/rest_entity.h"
+
+namespace http::rest_client {
+
+rest_entity::rest_entity(std::string_view root_url)
+  : _root_url{
+      root_url.ends_with("/") ? root_url : fmt::format("{}/", root_url)} {}
+
+request_builder rest_entity::create(
+  optional_headers headers, optional_query_params query_params) {
+    apply_optionals(headers, query_params);
+    return _rb.target(resource_url()).method(create_verb());
+}
+
+request_builder rest_entity::list(
+  optional_headers headers, optional_query_params query_params) {
+    apply_optionals(headers, query_params);
+    return _rb.target(resource_url()).method(retrieve_verb());
+}
+
+request_builder rest_entity::get(
+  ss::sstring identifier,
+  ss::sstring subpath,
+  optional_headers headers,
+  optional_query_params query_params) {
+    apply_optionals(headers, query_params);
+    auto target = fmt::format("{}/{}", resource_url(), identifier);
+    if (!subpath.empty()) {
+        target = fmt::format("{}/{}", target, subpath);
+    }
+    return _rb.target(target).method(retrieve_verb());
+}
+
+request_builder rest_entity::update(
+  ss::sstring identifier,
+  optional_headers headers,
+  optional_query_params query_params) {
+    apply_optionals(headers, query_params);
+    return _rb.target(fmt::format("{}/{}", resource_url(), identifier))
+      .method(update_verb());
+}
+
+request_builder rest_entity::delete_(
+  ss::sstring identifier,
+  optional_headers headers,
+  optional_query_params query_params) {
+    apply_optionals(headers, query_params);
+    return _rb.target(fmt::format("{}/{}", resource_url(), identifier))
+      .method(delete_verb());
+}
+
+rest_entity& rest_entity::with_host(std::string_view host) {
+    _rb.host(host);
+    return *this;
+}
+
+void rest_entity::apply_optionals(
+  optional_headers headers, optional_query_params query_params) {
+    if (headers.has_value()) {
+        _rb.headers(headers.value());
+    }
```

Comment:
My mental model for this class is a little off, but I'm curious why we're keeping track of a long-lived _rb. Are the arguments/headers/params always meant to be reused between different the different method calls? Or is the idea that a callers just use a `rest_entity` for just one request? If the latter, maybe it's worth r-value qualifying the create/list/get/... methods

Reply:
> I'm curious why we're keeping track of a long-lived _rb

Do you mean as opposed to building an http request object as early as possible? Initially that was what I tried but I realized that keeping the request builder around and passing it out of the methods of the rest entity (and building the actual request as late as possible) is a lot more flexible. Take the code here as an example:

https://github.com/redpanda-data/redpanda/blob/7ac93ccdfa1f06e583fc39fd44e6213b98688b48/src/v/iceberg/rest_client/catalog_client.cc#L177-L186

We set the host late in the process, as well as content size, because the rest entity does not know about the payload or the host (it could take these as c-tor args, but keeping this in the catalog client results in less code duplication). So the catalog client is able to set the host for all requests in one place and handle payload size, while the rest entity itself builds the initial request. 

> Or is the idea that a callers just use a rest_entity for just one request? 

Yes, the rest entity is a pretty lightweight object, almost just a collection of methods with polymorphism, it just sets up the initial request builder and then sends it to the caller. The r-value approach makes sense in this context, I will try to add it.

Reply:
Having thought a bit more about this, I agree with avoiding keeping behind state. 

I think the right thing to do here would be to remove request related state altogether, IE instead of keeping the request builder as a field in the class, construct one request builder for each rest API call and return it as a result of that call. 

Each rest API call is now independent, and patterns like create-if-not-exists can be supported like (paraphrasing the actual calls):

```
table t{""};
if (!co_await perform(t.get("t1"))) {
 co_await perform(t.create("t1"));
}
```